numba/np/ufunc/dufunc.py

```diff
@@ -174,6 +174,21 @@ def add(self, sig):
         args, return_type = sigutils.normalize_signature(sig)
         return self._compile_for_argtys(args, return_type)
 
+    def __call__(self, *args, **kws):
+        """
+        Allow any argument that has overridden __array_ufunc__ (NEP-18)
+        to take control of DUFunc.__call__.
+        """
+        default = numpy_support.np.ndarray.__array_ufunc__
+
+        for arg in args + tuple(kws.values()):
+            if getattr(type(arg), "__array_ufunc__", default) is not default:
+                output = arg.__array_ufunc__(self, "__call__", *args, **kws)
+                if output is not NotImplemented:
+                    return output
+        else:
+            return super().__call__(*args, **kws)
+
```

Comment:
@jpivarski - this PR implemented:
```
>>> import numba as nb
>>> @nb.vectorize()
... def _square(x):
...     return x * x
... 
>>> _square        
<numba._DUFunc '_square'>
```
we need to add similar functionality to a `CUDAUFuncDispatcher` that handles `ufunc`s on `CUDA` to allow the following:
```
>>> @nb.vectorize(
...     target="cuda",
... )
... def _square_cuda(x):
...     return x * x
... 
>>> _square_cuda
<numba.cuda.vectorizers.CUDAUFuncDispatcher object at 0x7259f82fa790>

```